src/sparse/impl/KokkosSparse_spgemm_mkl_impl.hpp

```diff
@@ -47,634 +47,337 @@
-               typename KernelHandle::nnz_lno_t n,
-               typename KernelHandle::nnz_lno_t k,
-               in_row_index_view_type row_mapA,
-               in_nonzero_index_view_type entriesA,
-               in_nonzero_value_view_type valuesA,
-
-               bool transposeA, bin_row_index_view_type row_mapB,
-               bin_nonzero_index_view_type entriesB,
-               bin_nonzero_value_view_type valuesB, bool transposeB,
-               cin_row_index_view_type row_mapC,
-               cin_nonzero_index_view_type entriesC,
-               cin_nonzero_value_view_type valuesC, bool verbose = false) {
-#ifdef KOKKOSKERNELS_ENABLE_TPL_MKL
+template <>
+KOKKOS_INLINE_FUNCTION void MKLSparseMatrix<double>::get(MKL_INT &rows,
+                                                         MKL_INT &cols,
+                                                         MKL_INT *&rows_start,
+                                                         MKL_INT *&columns,
+                                                         double *&values) {
+  sparse_index_base_t indexing;
+  MKL_INT *rows_end;
+  mkl_call(mkl_sparse_d_export_csr(mtx, &indexing, &rows, &cols, &rows_start,
+                                   &rows_end, &columns, &values),
+           "Failed to export matrix with mkl_sparse_s_export_csr()!");
+  if (SPARSE_INDEX_BASE_ZERO != indexing) {
+    throw std::runtime_error(
+        "Expected zero based indexing in exported MKL sparse matrix\n");
+    return;
+  }
+}
 
-  typedef typename KernelHandle::nnz_lno_t idx;
+template <typename KernelHandle, typename a_rowmap_view_type,
+          typename a_index_view_type, typename a_values_view_type,
+          typename b_rowmap_view_type, typename b_index_view_type,
+          typename b_values_view_type, typename c_rowmap_view_type,
+          typename c_index_view_type, typename c_values_view_type>
+class MKLApply {
+ public:
+  typedef typename KernelHandle::nnz_lno_t nnz_lno_t;
   typedef typename KernelHandle::size_type size_type;
-
-  typedef typename KernelHandle::HandleTempMemorySpace HandleTempMemorySpace;
-  typedef typename Kokkos::View<int *, HandleTempMemorySpace>
-      int_temp_work_view_t;
-
   typedef typename KernelHandle::nnz_scalar_t value_type;
-
+  typedef typename KernelHandle::HandleTempMemorySpace HandleTempMemorySpace;
   typedef typename KernelHandle::HandleExecSpace MyExecSpace;
-  /*
-      if (!(
-          (Kokkos::SpaceAccessibility<typename
-     Kokkos::HostSpace::execution_space, typename
-     device1::memory_space>::accessible) &&
-          (Kokkos::SpaceAccessibility<typename
-     Kokkos::HostSpace::execution_space, typename
-     device2::memory_space>::accessible) &&
-          (Kokkos::SpaceAccessibility<typename
-     Kokkos::HostSpace::execution_space, typename
-     device3::memory_space>::accessible) )
-          ){
-        throw std::runtime_error ("MEMORY IS NOT ALLOCATED IN HOST DEVICE for
-     MKL\n"); return;
-      }
-  */
```

Comment:
Should we use `host_mirror`/`deep_copy()` ?

Reply:
The exception here was not thrown, but it was expected by the unit test (`is_expected_to_fail = true`), which passed silently accepting MKL crashing on GPU memory spaces... I fixed that, so we can run now MKL on GPU data (e.g. with CUDA).